### ISSUE TITLE
Smoke-artifact evaluation and vLLM serve + query

### ISSUE THREAD

**Issue by kreuzhofer** [kreuzhofer/nebius-slurm-qwen38-demo#10]
## Question

Do evaluation and serving work on a real fine-tuned artifact — and does the merged checkpoint load in vLLM at all?

Uses the smoke adapter, not the real run: the artifact is meaningless for accuracy but perfectly good for exercising code. Run eval at `--num-examples 100`.

Evaluation side:

- `evaluate.sbatch` on 1 GPU completes a three-way run (base + merged smoke LoRA), and every label, filename and heading matches the models actually passed
- **The base model's exact-match rate lands near ~39%, not 2–3%.** Corrected from the recipe research (#3): on the architecturally identical Qwen3.6-27B, this same dataset measured base at 39%. The README's Provenance section cites 2–3%, but that is from the Qwen3-era 8B/32B ancestors on different hardware. A base rate near 2% here means the prompt or the normalisation is broken, not that the model is weak — which is the opposite of the conclusion the old figure would have led to.
- `</think>` stripping actually triggers, or provably never needs to
- Chart, JSON and Markdown all get written

Serving side:

- **`model_type` check first, before spending a job on it.** Per #3, `merge_lora.py` is expected to emit `model_type: qwen3_5_text`, and a sibling repo reports vLLM cannot load that. Both `transformers` 5.16.1 (`modeling_auto.py`) and vLLM 0.28.0 (`registry.py`) do register the type, so their failure looks like a version artifact — but this is the single biggest untested risk in the LoRA path and it is cheap to check. Read the merged `config.json` before launching the server.
- `serve.sbatch` starts the vLLM OpenAI server on the merged checkpoint at TP=1
- **`--language-model-only` stays** — confirmed real in vLLM 0.28.0 by #3, which resolves Known-risk #1. Note it is a no-op on a merged text-only checkpoint, and the code comment describing it as skipping the vision tower is wrong and needs correcting.
- `query.sh` gets a SQL answer back, and the served output is consistent with what `evaluate.py` produced for the same question — the strongest available check that training-time and serving-time prompts agree
- Record startup time and GPU memory

Do **not** chase TP > 1 here; that's out on the fog deliberately, and #3 found a documented silent TP=4 hang for this model family on other hardware.

**Comment by kreuzhofer** [comment]
## Resolution

Evaluation and serving both work on the smoke artifact. Committed `75f64c4`. Two results matter far more than "it ran".

### 1. The metric mostly measures formatting, not SQL — graduated to [#15](https://github.com/kreuzhofer/nebius-slurm-qwen38-lora-demo/issues/15)

Base vs the 25-step smoke adapter, 100 held-out examples:

```
Qwen3.8-27B (base) :   1.0%  (1/100)
qwen3.8-27b-sql    :  72.0%  (72/100)  +71.0pp
```

Base at **1.0%** is far below both the README's cited 2–3% and [#3](https://github.com/kreuzhofer/nebius-slurm-qwen38-lora-demo/issues/3)'s predicted ~39%. By this ticket's own criterion that means the prompt or the normalisation is broken — and it is the normalisation:

| normalisation | base | 25-step LoRA |
|---|---|---|
| as shipped | **1/100** | 72/100 |
| + strip ``` fences | 3/100 | 72/100 |
| + normalise quote style | 16/100 | 73/100 |
| **+ both** | **58/100** | **73/100** |
| answers containing a fence | **72/100** | 0/100 |

The base model writes largely correct SQL and is scored at 1% because 72% of its answers arrive in markdown fences and it prefers `'single'` quotes where the dataset uses `"double"`. The semantic gain from fine-tuning is **~58% → 73%, about +15pp**, not +71. This also explains why a 25-step adapter already scores 72%: format compliance is learned almost immediately, and that is what the metric chiefly rewards — and it very likely explains the predecessor repo's "2–3% base" figure as the same artifact.

`normalize_sql` was deliberately **not** changed: it is the definition of the published number, so the choice belongs to [#15](https://github.com/kreuzhofer/nebius-slurm-qwen38-lora-demo/issues/15), which now blocks the real run's publication. Documented in the README.

### 2. Serving works, and Known-risk #1 is closed on direct evidence

- vLLM 0.28.0 **accepted `--language-model-only`** — `'language_model_only': True` in the parsed non-default args, zero argument rejections, empty stderr — and accepted it against a text-only `qwen3_5_text` config, which was the open question. `--reasoning-parser qwen3` and `--default-chat-template-kwargs '{"enable_thinking": false}'` were accepted too.
- **The merged checkpoint serves.** This was flagged in [#3](https://github.com/kreuzhofer/nebius-slurm-qwen38-lora-demo/issues/3) as the biggest untested risk, on a sibling repo's report that vLLM cannot load `qwen3_5_text`. It loads fine here — so that report was a version artifact, as [#4](https://github.com/kreuzhofer/nebius-slurm-qwen38-lora-demo/issues/4) suspected.
- **Train-vs-serve prompts agree exactly.** Asked the server the same question `evaluate.py` scored, and it returned `SELECT venue FROM table_name_50 WHERE away_team = "essendon"` — byte-identical. The strongest available check that the shared `build_prompt` path holds through serving.
- `query.sh` unmodified returned correct SQL for its default question: `SELECT department, AVG(salary) FROM employees GROUP BY department`.

**Serving measurements:** model load **50.13 GiB in 25.5s**; `torch.compile` 35.5s; init engine total **301.6s**; GPU KV cache **1,899,098 tokens** (231.8x concurrency at 8192 tokens/request), which is unusually generous because only 16 of 64 layers cache at all. vLLM also logs `Setting attention block size to 784 tokens to ensure attention page size >= mamba page size` — the hybrid architecture reshaping KV paging, worth knowing before tuning `--max-model-len`.

Job 820 was cancelled after querying; `vllm serve` runs indefinitely and would otherwise hold a GPU.

### 3. `query.sh` defects — fixed, since this ticket owns whichever bite

The escaping bug **does** bite, demonstrated live: a schema containing `DEFAULT "x"` produced `400 json_invalid: Expecting ',' delimiter`. Fixed by building the body with `json.dumps`; verified it now survives double quotes and backslashes. Also fixed: `set -e` (this was the only script without it), `curl --fail-with-body -m 120`, an endpoint wait before querying (squeue's NODELIST is empty while PENDING, so the old code mis-diagnosed a still-loading server as "no job found"), and the undocumented 5th positional.

The system prompt is now **imported from `sft_common`** instead of being a third copy — which was the concrete way the README's "cannot drift apart" claim was false for serving.

### Documentation corrected in this ticket

`serve.sbatch`'s `--language-model-only` comment (wrong twice: it disables multimodal *inputs*, and it is a no-op on a merged checkpoint), its TP reasoning (24 query heads, not linear-key), the README Known-risks preamble, and the README metric discussion. Per instruction, doc fixes land with the discovery rather than being deferred.

### Not done here

The GDN kernel path is still unidentified — vLLM's logs describe its own kernels, not what `transformers` binds during training. Still needs instrumentation.
